File: src/core/Assignment.cpp
#include "wme/core/Assignment.hpp"

#include <algorithm>
#include <cmath>

namespace wme {
namespace {

// 최단경로 증강 방식 헝가리안.
// 내부적으로 정사각으로 패딩하지 않고, 행마다 한 번씩 증강경로를 찾는다.
// 잠재값(u, v)을 유지해 각 증강이 O(n*m) 이 되도록 한다.
struct Solver {
    std::size_t n, m;                 // n = rows, m = cols (n <= m 가정)
    const std::vector<double>& c;
    bool transposed;

    std::vector<double> u, v;         // 쌍대 잠재값
    std::vector<int>    match_col;    // 열 -> 행
    std::vector<int>    way;          // 교대 경로 복원용

    Solver(std::size_t rows, std::size_t cols, const std::vector<double>& cost, bool t)
        : n(rows), m(cols), c(cost), transposed(t),
          u(rows + 1, 0.0), v(cols + 1, 0.0),
          match_col(cols + 1, -1), way(cols + 1, 0) {}

    [[nodiscard]] double at(std::size_t row, std::size_t col) const {
        return transposed ? c[col * n + row] : c[row * m + col];
    }

    void run() {
        constexpr double kInf = std::numeric_limits<double>::infinity();

        // 행마다 새로 잡던 두 버퍼를 밖으로 뺀다. 행 시작마다 값만 다시 채우므로
        // 계산은 그대로이고 할당만 n 회에서 1 회로 줄어든다.
        std::vector<double> minv(m + 1, kInf);
        std::vector<char>   used(m + 1, 0);

        for (std::size_t i = 1; i <= n; ++i) {
            match_col[0] = static_cast<int>(i);
            std::size_t j0 = 0;

            std::fill(minv.begin(), minv.end(), kInf);
            std::fill(used.begin(), used.end(), static_cast<char>(0));

            do {
                used[j0] = 1;
                const auto i0 = static_cast<std::size_t>(match_col[j0]);
                double     delta = kInf;
                std::size_t j1 = 0;

                for (std::size_t j = 1; j <= m; ++j) {
                    if (used[j]) continue;
                    const double cur = at(i0 - 1, j - 1) - u[i0] - v[j];
                    if (cur < minv[j]) { minv[j] = cur; way[j] = static_cast<int>(j0); }
                    if (minv[j] < delta) { delta = minv[j]; j1 = j; }
                }
                if (!std::isfinite(delta)) break;   // 도달 가능한 열이 없다

                for (std::size_t j = 0; j <= m; ++j) {
                    if (used[j]) {
                        u[static_cast<std::size_t>(match_col[j])] += delta;
                        v[j] -= delta;
                    } else {
                        minv[j] -= delta;
                    }
                }
                j0 = j1;
            } while (match_col[j0] != -1);

            // 증강경로를 따라 매칭을 뒤집는다
            while (j0 != 0) {
                const auto j1 = static_cast<std::size_t>(way[j0]);
                match_col[j0] = match_col[j1];
                j0 = j1;
            }
        }
    }
};

}  // namespace

AssignmentResult solveAssignment(const std::vector<double>& cost,
                                 std::size_t rows, std::size_t cols) {
    AssignmentResult out;
    out.row_to_col.assign(rows, -1);
    out.col_to_row.assign(cols, -1);
    if (rows == 0 || cols == 0 || cost.size() != rows * cols) return out;

    // 알고리즘은 행 수 <= 열 수를 가정하므로 필요하면 전치해서 푼다
    const bool transposed = rows > cols;
    Solver solver(transposed ? cols : rows, transposed ? rows : cols, cost, transposed);
    solver.run();

    for (std::size_t j = 1; j <= solver.m; ++j) {
        const int i = solver.match_col[j];
        if (i <= 0) continue;

        const std::size_t r = transposed ? (j - 1) : static_cast<std::size_t>(i - 1);
        const std::size_t k = transposed ? static_cast<std::size_t>(i - 1) : (j - 1);
        if (r >= rows || k >= cols) continue;

        // 금지 쌍이 억지로 배정됐다면 버린다.
        // 큰 비용으로 막아도 다른 선택지가 없으면 알고리즘은 그 쌍을 고른다.
        const double pair_cost = cost[r * cols + k];
        if (pair_cost >= kInfeasible) continue;

        out.row_to_col[r] = static_cast<int>(k);
        out.col_to_row[k] = static_cast<int>(r);
        out.total_cost += pair_cost;
        ++out.matched;
    }
    return out;
}

}  // namespace wme

File: python/wme/reference/assignment.py
"""전역 최적 이분 할당 참조 구현.

C++ src/core/Assignment.cpp 와 같은 의미를 가져야 한다:
  - 직사각 행렬 지원, 남는 행/열은 -1
  - 비용이 INFEASIBLE 이상인 쌍은 절대 배정하지 않는다
    (게이트를 통과 못한 연관은 억지로 맺지 말고 새 토큰을 만들어야 한다)
"""

from __future__ import annotations

import numpy as np

INFEASIBLE = 1e18


def solve_assignment(cost: np.ndarray) -> tuple[np.ndarray, np.ndarray, float]:
    """헝가리안(최단경로 증강).

    Returns:
        row_to_col: (rows,) 배정된 열 인덱스, 미배정은 -1
        col_to_row: (cols,) 배정된 행 인덱스, 미배정은 -1
        total_cost: 배정된 쌍의 비용 합
    """
    cost = np.asarray(cost, dtype=float)
    if cost.ndim != 2 or cost.size == 0:
        rows = cost.shape[0] if cost.ndim == 2 else 0
        cols = cost.shape[1] if cost.ndim == 2 else 0
        return np.full(rows, -1, dtype=int), np.full(cols, -1, dtype=int), 0.0

    rows, cols = cost.shape
    transposed = rows > cols
    C = cost.T if transposed else cost
    n, m = C.shape                      # n <= m 보장

    u = np.zeros(n + 1)
    v = np.zeros(m + 1)
    match_col = np.full(m + 1, -1, dtype=int)   # 열 -> 행
    way = np.zeros(m + 1, dtype=int)

    for i in range(1, n + 1):
        match_col[0] = i
        j0 = 0
        minv = np.full(m + 1, np.inf)
        used = np.zeros(m + 1, dtype=bool)

        while True:
            used[j0] = True
            i0 = match_col[j0]
            delta = np.inf
            j1 = 0
            for j in range(1, m + 1):
                if used[j]:
                    continue
                cur = C[i0 - 1, j - 1] - u[i0] - v[j]
                if cur < minv[j]:
                    minv[j] = cur
                    way[j] = j0
                if minv[j] < delta:
                    delta = minv[j]
                    j1 = j
            if not np.isfinite(delta):
                break

            for j in range(m + 1):
                if used[j]:
                    u[match_col[j]] += delta
                    v[j] -= delta
                else:
                    minv[j] -= delta
            j0 = j1
            if match_col[j0] == -1:
                break

        # 증강경로를 따라 매칭을 뒤집는다
        while j0 != 0:
            j1 = way[j0]
            match_col[j0] = match_col[j1]
            j0 = j1

    row_to_col = np.full(rows, -1, dtype=int)
    col_to_row = np.full(cols, -1, dtype=int)
    total = 0.0

    for j in range(1, m + 1):
        i = match_col[j]
        if i <= 0:
            continue
        r = (j - 1) if transposed else (i - 1)
        k = (i - 1) if transposed else (j - 1)
        if r >= rows or k >= cols:
            continue
        # 금지 쌍이 억지로 배정됐다면 버린다
        if cost[r, k] >= INFEASIBLE:
            continue
        row_to_col[r] = k
        col_to_row[k] = r
        total += float(cost[r, k])

    return row_to_col, col_to_row, total
